Fix "Required parameter 'theme_dir' was not passed" error when emulating an environment inside an emulated area

dnsv opened the pull request:
### Description
Environment emulation inside an area emulation returns the error "Required parameter 'theme_dir' was not passed". The classes `\Magento\Theme\Model\View\Design` and `\Magento\Theme\Model\Theme` retrieve their area from the app state unless the area is emulated (see their `getArea()` method). When the environment is emulated inside an area emulation, the app state's area is used. This causes incorrect behaviour when the environment is being emulated.

### Manual testing scenarios

```php
use \Magento\Framework\App\Area;

class Emulation
{
    /**
     * @var \Magento\Store\Model\App\Emulation
     */
    protected $appEmulation;

    /**
     * @var \Magento\Framework\App\State
     */
    protected $appState;

    public function __construct(
        \Magento\Store\Model\App\Emulation $appEmulation,
        \Magento\Framework\App\State $appState
    ) {
        $this->appEmulation = $appEmulation;
        $this->appState = $appState;
    }

    public function execute()
    {
        $this->appState->emulateAreaCode(Area::AREA_ADMINHTML, function () {
            $this->appEmulation->startEnvironmentEmulation(1, Area::AREA_FRONTEND, true);
            $this->appEmulation->stopEnvironmentEmulation();
        });
    }
}
```

I've executed this class using a cronjob (the cache needs to be flushed before starting). I have a store with the id "1" (lets call this store "default"). The default store uses the theme `Magento/luma`. `startEnvironmentEmulation(..)` at some point tries to load the default store's theme. `\Magento\Theme\Model\Theme getFullPath()` gets invoked and returns `adminhtml/Magento/luma`. The theme's full path is wrong as the app state's area is being used instead of the environments area, because the app state is emulated. This then prevents the theme from being loaded and produces an error. I've added a method to also check if the environment is being emulated and used it to load the correct area. 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Fix "Required parameter 'theme_dir' was not passed" error when emulating an environment inside an emulated area

magento-engcom-team commented:
Hi @dnsv. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

VladimirZaets commented on app/code/Magento/Store/Model/App/Emulation.php:
```diff
@@ -183,6 +183,16 @@ public function stopEnvironmentEmulation()
+     *
+     * @return bool
+     */
+    public function isEnvironmentEmulated()
```
Hi @dnsv. Thanks for the collaboration, due to [Magento backward-compatible guide](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/) and [Magento Technical Guidelines](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html) we can create new public methods only in special cases where there are really needed. In the current case, please make this function `private`

dnsv replied:
Hi @VladimirZaets. `isEnvironmentEmulated()` is called in `\Magento\Theme\Model\Theme` and `\Magento\Theme\Model\View\Design`. The method therefore needs to be public.

slavvka replied:
@dnsv You are correct but this class is marked with `@api` that means that it cannot be changed (at least in patch releases). So please reimplement the solution as it proposed here https://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/#php. Refer to *Introduction of a method to a class or interface* section

VladimirZaets commented on app/code/Magento/Theme/Model/Theme.php:
```diff
@@ -113,6 +119,7 @@ public function __construct(
         \Magento\Framework\View\Design\Theme\ImageFactory $imageFactory,
         \Magento\Framework\View\Design\Theme\Validator $validator,
         \Magento\Framework\View\Design\Theme\CustomizationFactory $customizationFactory,
+        \Magento\Store\Model\App\Emulation $appEmulation,
```
Due to [Magento backward-compatible guide](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/) we can't add new required parameters to the constructor. For that we can use optional parameters, you can look example in the guide.

VladimirZaets commented on app/code/Magento/Theme/Model/Theme.php:
```diff
@@ -79,6 +79,11 @@ class Theme extends \Magento\Framework\Model\AbstractModel implements ThemeInter
+    /**
+     * @var \Magento\Store\Model\App\Emulation
+     */
+    protected $appEmulation;
```
According to [Magento technical guidelines](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html) this property should be private

VladimirZaets commented:
If you have any questions feel free to ask, thanks

Commit:
Update fix for emulating an environment inside an emulated area to be backward compatible and to follow technical guidelines

Commit:
Change class string variable to constant in DesignTest and add  appEmulation variable to ThemeTest

dnsv commented:
@VladimirZaets I've added the suggested changes. Could you please check if everything is ok now.

Commit:
Refactor app emulation changes to be backward compatible

dnsv commented:
@slavvka I've moved all the logic from `Magento\Store\Model\App\Emulation` to a new class `Magento\Store\Model\App\EnvironmentEmulation` and converted the old one to a proxy. I'm a bit unsure about one part. I’m calling the new class in the constructor as `$appEmulation`. But this approach may cause some backward compatibility issues as any di.xml argument injections aren’t being applied to the new class. Is there any preferred way to handle this? I’ve currently removed all old constructor parameters as they aren’t used in this class anymore and argument injections aren’t causing compilation errors.

Can the method `isEnvironmentEmulated()` be introduced to `Magento\Store\Model\App\Emulation` in 2.3? My initial solution should then probably be suited for a a forwardport.

slavvka commented:
Hey @dnsv. Sorry for not having answered you earlier. We don't care about BIC issues with `di.xml` since we don not recommend to inherit classes unless they are designed to be inherited (e.g. abstract). But you shouldn't have removed arguments from old class's constructor as we need to preserve signatures. Please revert that.

As for 2.3, the situation is the same as with 2.2 since 2.3 is already in release stage and cannot accept such changes. So you will need to up-port it as is.

Commit:
Re-add constructor parameters to Magento\Store\Model\App\Emulation

dnsv commented:
@slavvka Thanks for the info. I've re-added the constructor parameters.

slavvka commented on app/code/Magento/Store/Model/App/Emulation.php:
```diff
@@ -9,62 +9,22 @@
 /**
  * @api
  * @since 100.0.2
+ * @deprecated
```
Please mention here the reason of deprecation

slavvka commented on app/code/Magento/Store/Model/App/EmulationInterface.php:
```diff
@@ -0,0 +1,41 @@
+
+use Magento\Framework\App\Area;
+
+interface EmulationInterface
```
Please add here a description of the interface

slavvka commented on app/code/Magento/Store/Model/App/EnvironmentEmulation.php:
```diff
@@ -0,0 +1,241 @@
+ * See COPYING.txt for license details.
+ */
+
+/**
```
Please move the description to the class definition and add some details it if possible

slavvka commented on app/code/Magento/Store/Model/App/EnvironmentEmulation.php:
```diff
@@ -0,0 +1,241 @@
+
+use \Magento\Framework\App\Area;
+
+class EnvironmentEmulation extends \Magento\Framework\DataObject implements \Magento\Store\Model\App\EmulationInterface
```
Here :)

slavvka commented on app/code/Magento/Store/Model/App/EnvironmentEmulation.php:
```diff
@@ -0,0 +1,241 @@
+     *
+     * @return bool
+     */
+    public function isEnvironmentEmulated()
```
Why didn't you add this public method to the interface?

slavvka reviewed (changes requested):
@dnsv Thank you for fast reacting. Could you please also make requested changes?

Commit:
Add isEnvironmentEmulated() to the EmulationInterface, Update DocBlocks

slavvka commented:
@dnsv Thank you for the changes!

slavvka commented on app/code/Magento/Store/Model/App/Emulation.php:
```diff
@@ -4,67 +4,26 @@
     /**
-     * @var \Magento\Store\Model\StoreManagerInterface
-     */
-    protected $_storeManager;
```
Since it is an `@api` class you cannot remove protected/public methods, properties and constants. Please leave it public interface as is.

Commit:
Re-add protected properties to Store\Model\App\Emulation

slavvka commented:
Hey @dnsv Look, since this PR introduces BIC, we want it to be delivered to 2.3 firstly. Could you please create a PR to 2.3, we will process it and then merge this one?

dnsv commented:
@slavvka I'm on it.

sidolov commented:
Hi @dnsv , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

dnsv commented:
Hi @sidolov! I've submited the requested forwardport 9 days ago (#19287). This pull request should be done on my part. Should I leave it closed until the forwardport is approved and merged?

sidolov commented:
Hi @dnsv , sure, you can reopen it once PR to 2.3-dedelop will be merged
